### PR TITLE
Feat/cache queries

### PR DESCRIPTION
**Pull request by JensZack**
Cache results from _get_id
In the run of:
`r2x -i temp_data/oct1024_Con_NG  --input-model=reeds-US --output-model=plexos --weather-year=2012 -o .out/$CASETS --year 2035 --plugins pcm_defaults emission_cap break_gens imports --debug --upgrade --flags daily-budgets=true`

In this case:
_get_id is called 434408 times
there are only 5397 unique calls of _get_id

Caching _get_id results leads to the performance gains:
plexos_exporter: 47.3s -> 19.7s
_get_id cumulative: 28.8s -> 1.7s

### CODE REVIEW

**Comment by JensZack** [comment]
I would like to clean up the cacheing mechanism to only use class/object name, for id queries in the object table. This also raises questions about what uniqueness can be enforced in other tables where _get_id is used repeatedly.

File: src/plexosdb/sqlite.py [hunk 1]
```diff
@@ -459,14 +461,17 @@ def add_object(
...
+        try:
+            with self._conn as conn:
+                cursor = conn.cursor()
+                cursor.execute(
+                    f"INSERT INTO {Schema.Objects.name}(name, class_id, category_id, GUID, description) "
+                    f"VALUES({placeholders})",
+                    params,
+                )
+                object_id = cursor.lastrowid
+        except sqlite3.IntegrityError as err:
```
Comment (pesap) [inline src/plexosdb/sqlite.py:473]: I think it might be worth that we manually show what error it is.

File: tests/data/plexosdb.xml [hunk 2]
```diff
@@ -82,7 +82,7 @@
 	<t_object>
 		<object_id>3</object_id>
 		<class_id>2</class_id>
-		<name>SolarPV01</name>
+		<name>SolarPV02</name>
```
Comment (pesap) [inline tests/data/plexosdb.xml:85]: Is this object repeated already?

File: tests/test_plexosdb_sqlite.py [hunk 3]
```diff
@@ -156,27 +159,32 @@ def test_get_collection_id(db):
...
+    # Add generator with same name different category
     with pytest.raises(ValueError):
-        _ = db.get_object_id(gen_01_name, class_name=ClassEnum.Generator)
+        gen_01_name = "gen1"
+        gen_id = db.add_object(
+            gen_01_name,
+            ClassEnum.Generator,
+            CollectionEnum.Generators,
+            description="Test Gen",
+        )
```
Comment (pesap) [inline tests/test_plexosdb_sqlite.py:185]: This is what raises the `ValueError`, no? Maybe we can move line 180 before the `with` statement and remove the `db.get_object_id`

File: tests/test_plexosdb_sqlite.py [hunk 4]
```diff
@@ -12,8 +14,13 @@ def db_empty() -> "PlexosSQLite":
...
 @pytest.fixture
-def db(data_folder) -> PlexosSQLite:
-    return PlexosSQLite(xml_fname=data_folder.joinpath(DB_FILENAME))
+def db(data_folder) -> Generator[PlexosSQLite, None, None]:
+    xml_fname = data_folder.joinpath(DB_FILENAME)
+    xml_copy = data_folder.joinpath(f"copy_{DB_FILENAME}")
+    shutil.copy(xml_fname, xml_copy)
+    db = PlexosSQLite(xml_fname=xml_copy)
+    yield db
+    xml_copy.unlink()
```
Comment (pesap) [inline tests/test_plexosdb_sqlite.py:23]: can we create the copy on a `tmp_path` instead of the data folder?

**Review by pesap — CHANGES_REQUESTED** [review]
LGTM. We can merge once the comments are addressed.

File: tests/test_plexosdb_sqlite.py (hunk 4 above)
Comment (JensZack) [reply tests/test_plexosdb_sqlite.py:23]: oh yeah, that makes sense. I was trying to use with TemporaryFile, but it was not deleting the file when the tests completed, I can try the directory approach

File: src/plexosdb/sqlite.py (hunk 1 above)
Comment (JensZack) [reply src/plexosdb/sqlite.py:473]: I was trying to maintain the ValueError for consistency with previous get_object behavior. I also like the idea that a module using plexosdb could catch an error from PlexosDB.add_object without importing sqlite3. I am happy to remove this error conversion though.

**Review by pesap — COMMENTED** [review]
Can you also improve the coverage to at least 92%? I think we are missing like 7 lines of coverage. I think it is from the cache implementation however the CI is not showing where. We will need to run `coverage` command locally.

File: tests/test_plexosdb_sqlite.py (hunk 3 above)
Comment (JensZack) [reply tests/test_plexosdb_sqlite.py:187]: Done

File: tests/test_plexosdb_sqlite.py (hunk 4 above)
Comment (JensZack) [reply tests/test_plexosdb_sqlite.py]: Done

File: tests/test_plexosdb_sqlite.py [hunk 5]
```diff
@@ -174,17 +174,14 @@ def test_get_object_id(db):
     gen_id_get = db.get_object_id(gen_01_name, class_name=ClassEnum.Generator)
     assert gen_id == gen_id_get
 
-    # Add generator with same name different category
+    # Add generator with same name and no category
```
[the remark this answers is not in the document]
Comment (JensZack) [reply tests/test_plexosdb_sqlite.py:180]: This is just a comment about the test, not a todo

File: tests/test_plexosdb_sqlite.py (hunk 3 above)
Comment (JensZack) [reply tests/test_plexosdb_sqlite.py:188]: I think I fixed this

File: tests/test_plexosdb_sqlite.py [hunk 6]
```diff
@@ -12,8 +17,14 @@ def db_empty() -> "PlexosSQLite":
...
 
 @pytest.fixture
-def db(data_folder) -> PlexosSQLite:
-    return PlexosSQLite(xml_fname=data_folder.joinpath(DB_FILENAME))
+def db(data_folder: Path) -> Generator[PlexosSQLite, None, None]:
+    tmp_dir = TemporaryDirectory()  # dir=data_folder.parent)
+    xml_fname = data_folder / DB_FILENAME
+    xml_copy = Path(tmp_dir.name) / f"copy_{DB_FILENAME}"
+    shutil.copy(xml_fname, xml_copy)
+    db = PlexosSQLite(xml_fname=str(xml_copy))
```
Comment (pesap) [inline tests/test_plexosdb_sqlite.py:25]: We do not need the TemporaryDirectory for this, we can just reuse the temporary file created by pytest. See: https://docs.pytest.org/en/stable/how-to/tmp_path.html

```suggestion
def db(data_folder: Path, tmp_path) -> Generator[PlexosSQLite, None, None]:
    xml_fname = tmp_path / DB_FILENAME
    xml_copy = Path(tmp_path) / f"copy_{DB_FILENAME}"
    shutil.copy(xml_fname, xml_copy)
    db = PlexosSQLite(xml_fname=str(xml_copy))
```

**Review by pesap — CHANGES_REQUESTED** [review]
Can you run the coverage with this command locally:

```
coverage run -m pytest -vvl && coverage report -m --sort
```

**Review by pesap — APPROVED** [review]

**Pull request merged**